tests/components/zha/test_cover.py

```diff
@@ -129,3 +158,151 @@ async def get_chan_attr(*args, **kwargs):
+
+    # set position UI command fails
+    with patch("zigpy.zcl.Cluster.request", side_effect=asyncio.TimeoutError):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_SET_COVER_POSITION,
+            {"entity_id": entity_id, "position": 47},
+            blocking=True,
+        )
+        assert cluster_level.request.call_count == 1
+        assert cluster_level.request.call_args[0][0] is False
+        assert cluster_level.request.call_args[0][1] == 0x0004
+        assert int(cluster_level.request.call_args[0][3] * 100 / 255) == 47
+        assert hass.states.get(entity_id).attributes[ATTR_CURRENT_POSITION] == 0
+
+    # set position UI success
+    with patch(
+        "zigpy.zcl.Cluster.request", AsyncMock(return_value=[0x5, zcl_f.Status.SUCCESS])
+    ):
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_SET_COVER_POSITION,
+            {"entity_id": entity_id, "position": 47},
+            blocking=True,
+        )
+        assert cluster_level.request.call_count == 1
+        assert cluster_level.request.call_args[0][0] is False
+        assert cluster_level.request.call_args[0][1] == 0x0004
+        assert int(cluster_level.request.call_args[0][3] * 100 / 255) == 47
+        assert hass.states.get(entity_id).attributes[ATTR_CURRENT_POSITION] == 47
+
+    # report position change
+    await send_attributes_report(hass, cluster_level, {8: 0, 0: 100, 1: 1})
+    assert hass.states.get(entity_id).attributes[ATTR_CURRENT_POSITION] == int(
+        100 * 100 / 255
+    )
+
+    # test rejoin
+    await async_test_rejoin(
+        hass, zigpy_shade_device, [cluster_level, cluster_on_off], (1,)
+    )
+    assert hass.states.get(entity_id).state == STATE_OPEN
+
+    # test cover stop
+    with patch("zigpy.zcl.Cluster.request", side_effect=asyncio.TimeoutError):
+        await hass.services.async_call(
+            DOMAIN, SERVICE_STOP_COVER, {"entity_id": entity_id}, blocking=True,
+        )
+        assert cluster_level.request.call_count == 1
+        assert cluster_level.request.call_args[0][0] is False
+        assert cluster_level.request.call_args[0][1] in (0x0003, 0x0007)
+
+
+async def test_restore_state(hass, zha_device_restored, zigpy_shade_device):
+    """Ensure states are restored on startup."""
+
+    mock_restore_cache(
+        hass,
+        (
+            State(
+                "cover.fakemanufacturer_fakemodel_e769900a_level_on_off_shade",
+                STATE_OPEN,
+                {ATTR_CURRENT_POSITION: 50},
+            ),
+        ),
+    )
+
+    hass.state = CoreState.starting
+
+    zha_device = await zha_device_restored(zigpy_shade_device)
+    entity_id = await find_entity_id(DOMAIN, zha_device, hass)
+    assert entity_id is not None
+
+    # test that the cover was created and that it is unavailable
+    assert hass.states.get(entity_id).state == STATE_OPEN
+    assert hass.states.get(entity_id).attributes[ATTR_CURRENT_POSITION] == 50
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5575421)